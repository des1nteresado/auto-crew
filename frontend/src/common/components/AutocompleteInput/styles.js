const styles = (theme) => ({
  autocompleteInput: {
    '& .MuiInputBase-input': {
      fontSize: 13,
      color: theme.palette.black,
    },
  },
  autocompleteList: {
    fontSize: 13,
    color: theme.palette.black,
  },
});

export default styles;
