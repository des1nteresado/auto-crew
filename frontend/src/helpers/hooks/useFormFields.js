import { useState } from 'react';

export const useFormFields = (initialState) => {
  const [fields, setValue] = useState(initialState);

  return [
    fields,
    (event) => {
      if (event.target.type === 'checkbox') {
        setValue({
          ...fields,
          [event.target.name]: event.target.checked,
        });
      } else {
        setValue({
          ...fields,
          [event.target.name]: event.target.value,
        });
      }
    },
  ];
};
