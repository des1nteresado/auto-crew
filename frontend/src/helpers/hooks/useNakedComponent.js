export const useNakedComponent = (component) => {
  return component?.type?.Naked;
};
