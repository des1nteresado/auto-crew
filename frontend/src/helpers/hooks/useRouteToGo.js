import { useCallback } from 'react';
import { useHistory } from 'react-router-dom';

export const useRouteToGo = (path) => {
  const history = useHistory();

  return useCallback(() => history.push(path), [path, history]);
};
