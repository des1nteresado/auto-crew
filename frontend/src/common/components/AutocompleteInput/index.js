import React, { useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import Autocomplete from '@material-ui/lab/Autocomplete';
import { TextField, withStyles } from '@material-ui/core';

import styles from './styles.js';

const AutocompleteInput = ({
  autocompleteValue,
  inputValue,
  valuesList,
  getOptionLabel,
  onAutocompleteChange,
  onInputChange,
  name,
  error,
  classes,
  helperText,
  label,
}) => {
  const [isTouched, setIsTouched] = useState(false);

  const onBlur = useCallback(() => {
    setIsTouched(true);
  }, [setIsTouched]);

  return (
    <Autocomplete
      value={autocompleteValue}
      classes={{
        paper: classes.autocompleteList,
      }}
      options={valuesList}
      getOptionLabel={getOptionLabel}
      onChange={(event, value) => onAutocompleteChange(value)}
      renderInput={(params) => (
        <TextField
          {...params}
          value={inputValue}
          classes={{
            root: classes.autocompleteInput,
          }}
          onChange={onInputChange}
          name={name}
          onBlur={onBlur}
          label={label}
          variant="outlined"
          error={isTouched && error}
          helperText={isTouched && error && helperText}
        />
      )}
    />
  );
};

AutocompleteInput.defaultProps = {
  inputValue: '',
  autocompleteValue: {},
  valuesList: [],
  label: '',
  style: '',
  helperText: '',
  type: 'text',
  error: false,
  onAutocompleteChange: () => {},
  onInputChange: () => {},
  getOptionLabel: () => {},
};

AutocompleteInput.propTypes = {
  classes: PropTypes.oneOfType([PropTypes.object]).isRequired,
  autocompleteValue: PropTypes.oneOfType([PropTypes.object]),
  valuesList: PropTypes.oneOfType([PropTypes.array]),
  onAutocompleteChange: PropTypes.func,
  onInputChange: PropTypes.func,
  inputValue: PropTypes.string,
  getOptionLabel: PropTypes.func,
  name: PropTypes.string.isRequired,
  helperText: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  label: PropTypes.string,
  error: PropTypes.bool,
  style: PropTypes.string,
  type: PropTypes.string,
};

export default React.memo(withStyles(styles)(AutocompleteInput));
