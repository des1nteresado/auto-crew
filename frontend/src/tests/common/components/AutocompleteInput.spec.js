import React from 'react';

import AutocompleteInput from 'common/components/AutocompleteInput';
import { useNakedComponent } from 'helpers/hooks/useNakedComponent';

const props = {
  name: 'test name',
};

const setUp = (props) => shallow(<AutocompleteInput {...props} />);

describe('AutocompleteInput', () => {
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation((init) => [init, setState]);

  let component;

  beforeEach(() => {
    component = setUp(props);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders AutocompleteInput component correctly', () => {
    expect(component).toMatchSnapshot();
  });

  // @TODO: figure out why it isn't working
  it('updates state correctly when input blur callback is calling', () => {
    const component = mount(<AutocompleteInput {...props} />);
    const input = component.find('input');

    expect(input.props().name).toBe('test name');
    input.simulate('blur');
    expect(setState).toHaveBeenCalledWith(true);
  });

  describe('default props', () => {
    const nakedComponent = useNakedComponent(AutocompleteInput);

    it('should use default onAutocompleteChange', () => {
      const result = nakedComponent.defaultProps.onAutocompleteChange();
      expect(result).toBe(undefined);
    });

    it('should use default onInputChange', () => {
      const result = nakedComponent.defaultProps.onInputChange();
      expect(result).toBe(undefined);
    });

    it('should use default getOptionLabel', () => {
      const result = nakedComponent.defaultProps.getOptionLabel();
      expect(result).toBe('');
    });
  });
});
