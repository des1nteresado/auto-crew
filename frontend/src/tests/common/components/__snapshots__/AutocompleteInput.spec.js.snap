// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocompleteInput renders AutocompleteInput component correctly 1`] = `
<AutocompleteInput
  autocompleteValue={Object {}}
  classes={
    Object {
      "autocompleteInput": "AutocompleteInput-autocompleteInput-1",
      "autocompleteList": "AutocompleteInput-autocompleteList-2",
    }
  }
  error={true}
  getOptionLabel={[Function]}
  helperText="error text"
  inputValue=""
  label=""
  name="name"
  onAutocompleteChange={[MockFunction]}
  onInputChange={[Function]}
  style=""
  type="text"
  valuesList={Array []}
/>
`;
